src/main/java/com/google/firebase/database/util/JsonMapper.java

```diff
@@ -46,93 +46,90 @@ public static String serializeJsonValue(Object object) throws IOException {
     if (object == null) {
       return "null";
     } else if (object instanceof String) {
-      return JSONObject.quote((String) object);
+      return GSON.toJson(object);
     } else if (object instanceof Number) {
       try {
-        return JSONObject.numberToString((Number) object);
-      } catch (JSONException e) {
+        return GSON.toJson(object);
+      } catch (Exception e) {
         throw new IOException("Could not serialize number", e);
       }
     } else if (object instanceof Boolean) {
       return ((Boolean) object) ? "true" : "false";
     } else {
-      try {
-        JSONStringer stringer = new JSONStringer();
-        serializeJsonValue(object, stringer);
-        return stringer.toString();
-      } catch (JSONException e) {
-        throw new IOException("Failed to serialize JSON", e);
-      }
-    }
-  }
-
-  private static void serializeJsonValue(Object object, JSONStringer stringer)
-      throws IOException, JSONException {
-    if (object instanceof Map) {
-      stringer.object();
-      @SuppressWarnings("unchecked")
-      Map<String, Object> map = (Map<String, Object>) object;
-      for (Map.Entry<String, Object> entry : map.entrySet()) {
-        stringer.key(entry.getKey());
-        serializeJsonValue(entry.getValue(), stringer);
-      }
-      stringer.endObject();
-    } else if (object instanceof Collection) {
-      Collection<?> collection = (Collection<?>) object;
-      stringer.array();
-      for (Object entry : collection) {
-        serializeJsonValue(entry, stringer);
-      }
-      stringer.endArray();
-    } else {
-      stringer.value(object);
+      return GSON.toJson(object);
     }
   }
 
+  @SuppressWarnings("unchecked")
   public static Map<String, Object> parseJson(String json) throws IOException {
     try {
-      return unwrapJsonObject(new JSONObject(json));
-    } catch (JSONException e) {
+      JsonReader jsonReader = new JsonReader(new StringReader(json));
+      return (Map) unwrapJson(jsonReader);
```

Comment:
I have reworked exception handling a bit. Instead of casting I just call unwrapJsonObject. But I have also add IllegalStateException exception because beginArray, beginObject (in gson code) internally throws in case there is unexpected token.